src/main/java/appeng/thirdparty/codechicken/lib/model/Quad.java

```diff
@@ -0,0 +1,557 @@
+	}
+
+	@Override
+	public void setApplyDiffuseLighting( boolean diffuse )
+	{
+		this.diffuseLighting = diffuse;
+	}
+
+	@Override
+	public void setTexture( TextureAtlasSprite texture )
+	{
+		this.sprite = texture;
+	}
+	//@formatter:on
+
+	@Override
+	public void put( int element, float... data )
+	{
+		if( full )
+		{
+			throw new RuntimeException( "Unable to add data when full." );
+		}
+		Vertex v = vertices[vertexIndex];
+		if( v == null )
+		{
+			v = new Vertex( format );
+			vertices[vertexIndex] = v;
+		}
+		System.arraycopy( data, 0, v.raw[element], 0, data.length );
+		if( element == ( format.elementCount - 1 ) )
+		{
+			vertexIndex++;
+			if( vertexIndex == 4 )
+			{
+				vertexIndex = 0;
+				full = true;
+			}
+		}
+	}
+
+	@Override
+	public void put( Quad quad )
+	{
+		copyFrom( quad );
+	}
+
+	@Override
+	public void pipe( IVertexConsumer consumer )
+	{
+		if( consumer instanceof ISmartVertexConsumer )
+		{
+			( (ISmartVertexConsumer) consumer ).put( this );
+		}
+		else
+		{
+			consumer.setQuadTint( tintIndex );
+			consumer.setQuadOrientation( orientation );
+			consumer.setApplyDiffuseLighting( diffuseLighting );
+			consumer.setTexture( sprite );
+			for( Vertex v : vertices )
+			{
+				for( int e = 0; e < format.elementCount; e++ )
+				{
+					consumer.put( e, v.raw[e] );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Used to reset the interpolation values inside the provided helper.
+	 *
+	 * @param helper The helper.
+	 * @param s The axis. side >> 1;
+	 *
+	 * @return The same helper.
+	 */
+	public InterpHelper resetInterp( InterpHelper helper, int s )
+	{
+		helper.reset(//
+				vertices[0].dx( s ), vertices[0].dy( s ),//
+				vertices[1].dx( s ), vertices[1].dy( s ),//
+				vertices[2].dx( s ), vertices[2].dy( s ),//
+				vertices[3].dx( s ), vertices[3].dy( s ) );
+		return helper;
+	}
+
+	/**
+	 * Clamps the Quad inside the box.
+	 *
+	 * @param bb The box.
+	 */
+	public void clamp( AxisAlignedBB bb )
+	{
+		for( Vertex vertex : vertices )
+		{
+			float[] vec = vertex.vec;
+			vec[0] = (float) MathHelper.clamp( vec[0], bb.minX, bb.maxX );
+			vec[1] = (float) MathHelper.clamp( vec[1], bb.minY, bb.maxY );
+			vec[2] = (float) MathHelper.clamp( vec[2], bb.minZ, bb.maxZ );
+		}
+
+		v1.set( vertices[3].vec );
+		t.set( vertices[1].vec );
+		v1.sub( t );
+
+		v2.set( vertices[2].vec );
+		t.set( vertices[0].vec );
+		v2.sub( t );
+
+		normal.cross( v2, v1 );
+		normal.normalize();
+
+		if( format.hasNormal )
+		{
+			for( Vertex vertex : vertices )
+			{
+				vertex.normal[0] = normal.x;
+				vertex.normal[1] = normal.y;
+				vertex.normal[2] = normal.z;
+				vertex.normal[3] = 0;
+			}
+		}
+		orientation = EnumFacing.getFacingFromVector( normal.x, normal.y, normal.z );
+	}
+
+	/**
+	 * Used to create a new quad complete copy of this one.
+	 *
+	 * @return The new quad.
+	 */
+	public Quad copy()
+	{
+		if( !full )
+		{
+			throw new RuntimeException( "Only copying full quads is supported." );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS00112)